crates/nu-command/src/filters/default.rs

```diff
@@ -110,70 +139,150 @@ fn default(
         || (default_when_empty
             && matches!(input, PipelineData::Value(ref value, _) if value.is_empty()))
     {
-        Ok(value.into_pipeline_data())
+        default_value.pipeline_data()
     } else if default_when_empty && matches!(input, PipelineData::ListStream(..)) {
         let PipelineData::ListStream(ls, metadata) = input else {
             unreachable!()
         };
         let span = ls.span();
         let mut stream = ls.into_inner().peekable();
         if stream.peek().is_none() {
-            return Ok(value.into_pipeline_data());
+            return default_value.pipeline_data();
         }
 
         // stream's internal state already preserves the original signals config, so if this
         // Signals::empty list stream gets interrupted it will be caught by the underlying iterator
         let ls = ListStream::new(stream, span, Signals::empty());
         Ok(PipelineData::ListStream(ls, metadata))
+    // Otherwise, return the input as is
     } else {
         Ok(input)
     }
 }
 
+/// A wrapper around the default value to handle closures and caching values
+enum DefaultValue<'a> {
+    Uncalculated(&'a EngineState, &'a Stack, Spanned<Closure>),
+    Calculated(Value),
+}
+
+impl<'a> DefaultValue<'a> {
+    fn new(engine_state: &'a EngineState, stack: &'a Stack, value: Value) -> Self {
+        let span = value.span();
+        match value {
+            Value::Closure { val, .. } => {
+                DefaultValue::Uncalculated(engine_state, stack, (*val).into_spanned(span))
+            }
+            _ => DefaultValue::Calculated(value),
+        }
+    }
+
+    fn value(&mut self) -> Result<Value, ShellError> {
+        match self {
+            DefaultValue::Uncalculated(engine_state, stack, closure) => {
+                let closure_eval = ClosureEvalOnce::new(engine_state, stack, closure.item.clone());
+                let value = closure_eval
+                    .run_with_input(PipelineData::Empty)?
+                    .into_value(closure.span)?;
+                *self = DefaultValue::Calculated(value.clone());
+                Ok(value)
+            }
+            DefaultValue::Calculated(value) => Ok(value.clone()),
+        }
+    }
+
+    fn pipeline_data(&mut self) -> Result<PipelineData, ShellError> {
+        self.value().map(|x| x.into_pipeline_data())
+    }
+}
+
+/// Given a record, fill missing columns with a default value
+fn fill_record(
+    record: Spanned<&mut Record>,
+    default_value: &mut DefaultValue,
+    columns: &[String],
+    empty: bool,
+) -> Result<Value, ShellError> {
+    for col in columns {
+        if let Some(val) = record.item.get_mut(col) {
+            if matches!(val, Value::Nothing { .. }) || (empty && val.is_empty()) {
+                *val = default_value.value()?;
+            }
+        } else {
+            record.item.push(col.clone(), default_value.value()?);
+        }
+    }
+    Ok(Value::record(record.item.clone(), record.span))
+}
```

Comment:
```suggestion
fn fill_record(
    mut record: Record,
    span: Span,
    default_value: &mut DefaultValue,
    columns: &[String],
    empty: bool,
) -> Result<Value, ShellError> {
    for col in columns {
        if let Some(val) = record.get_mut(col) {
            if matches!(val, Value::Nothing { .. }) || (empty && val.is_empty()) {
                *val = default_value.value()?;
            }
        } else {
            record.push(col.clone(), default_value.value()?);
        }
    }
    Ok(Value::record(record, span))
}
```

This just changes the function to operate on the `Record` value directly rather than working with a mutable reference. Previously, `record.item.clone()` is always cloned (so passing by mutable reference isn't really helping). We can avoid that clone by operating on the value directly.

To get a bit into the weeds, we *might* actually clone above when we call `into_record` on the `Value` containing the Record, as created an owned version of `Record` means we might have to pull it out of the `SharedCow`, but a "maybe clone" is better than an "always clone" 😄